### ISSUE TITLE
Move classes to the classes directory

### ISSUE THREAD

**Issue by azito122** [CLAMP-IT/moodle-blocks_filtered_course_list#107]
Especially locallib, which is where the individual filter classes live, is a bit overstuffed, and it isn't obvious that that is where those classes should be. I believe this is the exact problem the classes directory is intended to solve. I envision a file tree like:
```
classes/
|── privacy/
│   ├── bootstrap.css
├── filters/
│   ├── base.php (configline class)
│   ├── regex.php (regex_configline class)
│   ├── shortname.php (shortname_configline class)
...
...
```

Additionally, it might be slightly neater to use actual namespaces rather than long underscored classnames.

**Comment by kwiliarty** [comment]
I'm definitely open to a PR for something like this, and it is potentially a prerequisite to making filters pluggable. 

Here is something I would like to consider, though: Moodle clearly has some expectations about plugin structure. Directories like `lang`, `pix`, `tests` and `db` must be organized a certain way, and now with the Privacy API `classes/privacy` is part of that picture. `classes/event` has been that way for a while. Can you find any documentation about expectations for the classes directory? I woudn't want to mess with something that is reserved. Or can you check out how exemplary core modules do it? Is `classes/filters` a better choice than `filters` or `lib/filters`? I genuinely don't know if there is a principled reason to chose one option over another. 

Bottom line though: moving filter classes into separate files within a filters directory is a good idea.

**Comment by azito122** [comment]
That's a very good question. I don't know that there is a perfectly clear answer, but here are some quotes from relevant documentation:

From https://docs.moodle.org/dev/Coding_style#Namespaces:
> Namespaces are not required for any new code and there is no requirement to move non-namespaced classes to namespaces. If namespaces are used - they must conform to these rules.
>
> **Classes belonging to a namespace must be created in a classes directory inside a plugin** (e.g. mod/forum/classes), or for core code, in lib/classes or subsystemdir/classes.
> 
> The classname and filename for all namespaced classes must conform to the automatic class loading rules.

From https://docs.moodle.org/dev/NEWMODULE_Documentation#lib.php:
> All the newmodule specific functions, needed to implement the rest of the module logic should live in autoloaded classes inside the newmodule/classes directory. 

From https://docs.moodle.org/dev/Automatic_class_loading:
> Class file names must be class name without Frankenstyle prefix
> - e.g. Class mod_forum_some_class is stored in file mod/forum/classes/some_class.php
> - e.g. Class core_frankenstyle class is stored in lib/classes/frankenstyle.php

Additionally, while most modules seem to have just `/classes/event` or `/classes/privacy`, some do include other files. For example, the new `dataprivacy` admin tool includes a rather large number of custom classes, some even in subfolders (see https://github.com/moodlehq/moodle-tool_dataprivacy/tree/MOODLE_34_STABLE/classes).

All that being said, I have no particular preference for using `/classes` over `/filter` or some other custom folder. To my mind, the important thing is that classes themselves be well organized.

**Comment by kwiliarty** [comment]
Thanks for digging into it. That is very helpful information, and it reassures me to give you a green light to work on this. I'll accept it and assign it to you.
